NuclearData/ceNeutronData/aceDatabase/aceNeutronDatabase_class.f90

```diff
@@ -771,13 +783,122 @@ subroutine activate(self, activeMat)
 
     ! Configure Cache
     if (self % hasUrr) then
-      call cache_init(size( self % materials), size(self % nuclides), 1, maxval(self % nucToZaid))
+      call cache_init(size(self % materials), size(self % nuclides), 1, maxval(self % nucToZaid))
     else
-      call cache_init(size( self % materials), size(self % nuclides))
+      call cache_init(size(self % materials), size(self % nuclides))
     end if
 
   end subroutine activate
 
+  !!
+  !! Precomputes majorant cross section
+  !!
+  !! See nuclearDatabase documentation for details
+  !!
+  subroutine initMajorant(self, rand, silent)
+    class(aceNeutronDatabase), intent(inout) :: self
+    class(RNG), intent(inout)                :: rand
+    logical(defBool), intent(in), optional   :: silent
+    logical(defBool)                         :: loud
+    real(defReal), dimension(:), allocatable :: tmpGrid
+    integer(shortInt)                        :: i, j, matIdx, nNuc, nucIdx, isDone
+    type(intMap)                             :: nucMap
+    real(defReal)                            :: E, maj
+    integer(shortInt), parameter :: IN_SET = 1, NOT_PRESENT = 0
+
+    ! Set build console output flag
+    if(present(silent)) then
+      loud = .not.silent
+    else
+      loud = .true.
+    end if
+
+    if (loud) print '(A)', 'Building unionised energy grid'
+
+    ! Initialise energy grid
+    matIdx = self % activeMat(1)
+    nucIdx = self % materials(matIdx) % nuclides(1)
+    tmpGrid = self % nuclides(nucIdx) % eGrid
+
+    ! Add first nuclide to map to indicate it's been added
+    call nucMap % add(nucIdx, IN_SET)
+
+    ! Loop over active materials
+    do i = 1, size(self % activeMat)
+
+      ! Get current material index and number of nuclides in that material
+      matIdx = self % activeMat(i)
+      nNuc = size(self % materials(matIdx) % nuclides)
+
+      ! Loop over nuclides present in that material
+      do j = 1, nNuc
+
+        ! Get index and check if it's already been added to the grid
+        nucIdx = self % materials(matIdx) % nuclides(j)
+        isDone = nucMap % getOrDefault(nucIdx, NOT_PRESENT)
+
+        ! If it's a new nuclide, add its energy grid to the unionised one
+        if (isDone /= IN_SET) then
+
+          tmpGrid = concatenate(tmpGrid, self % nuclides(nucIdx) % eGrid)
+
+          ! Sort and remove duplicates
+          call quickSort(tmpGrid)
+          tmpGrid = removeDuplicatesSorted(tmpGrid)
+
+          ! Add nuclide to the set
+          call nucMap % add(nucIdx, IN_SET)
+
+        end if
+
+      end do
+    end do
```

Comment:
It is perhaps not at this moment but the (relatively) low hanging fruits for optimisation here would be to create a `real` version of the `dynArray_class` and use it as the `tmpGrid`. I guess it could take care of a large portion of the runtime by avoiding a lot of reallocation. 

Of course `dynArray` would than also have in-place versions of `removeDuplicates` and `quickSort` added.

Reply:
It is a bit involved, so I will just open an issue. It will be needed for actual use with burnup or so.